LFS-654: Better UI for Questionnaire wizard; includes: LFS-673, LFS-686, LFS-643

veronikaslc opened the pull request:
[LFS-654](https://phenotips.atlassian.net/browse/LFS-654): Better UI for Questionnaire wizard
[LFS-673](https://phenotips.atlassian.net/browse/LFS-673):The questionnaire wizard fails to show updated questionnaire requiredSubjectTypes property that was previously unset
[LFS-686](https://phenotips.atlassian.net/browse/LFS-686): In the questionnaire wizard, the last predefined answer option provided for a question does not get saved
[LFS-643](https://phenotips.atlassian.net/browse/LFS-643): Misplaced Data Type label in Admin's Edit Question dialog

Commit:
LFS-654: Better UI for Questionnaire wizard
First draft of better UI

Commit:
LFS-654: Better UI for Questionnaire wizard
Removed unused function

marta- reviewed (commented):
Attempting to edit a section appears to have no effect (changes are not saved) and results in a console error:
```
webpack:///./node_mo….development.js?:88 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in Section (created by Questionnaire)
    in div (created by ForwardRef(Grid))
    in ForwardRef(Grid) (created by WithStyles(ForwardRef(Grid)))
    in WithStyles(ForwardRef(Grid)) (created by Questionnaire)
```

marta- committed:
LFS-654: Better UI for Questionnaire wizard
Some cosmetic changes for a more user-friendly questionnaire layout: improved spacing and indentation, added avatars and to question and section cards for improved clarity.
The code is in draft mode and needs more cleanup. To do:
- extract inline styles and move to the style jsx,
- extract all comon functionality of Section and Question cards into a single reusable QuestionnaireItemCard component
- improve the display of question fields
- show all section properties in view mode

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/EditDialog.jsx:
```diff
@@ -162,6 +162,19 @@ let EditDialog = (props) => {
           <DialogContent>
             { !isEdit && titleField() }
             <Fields data={isEdit && data || {}} JSON={json[0]} edit={true} />
+            { data && type === 'Section' &&
+              <React.Fragment key={`${data.condition['@path']}`}>
+                <Typography>Condition:</Typography>
+                <input type='hidden' name={`${data['@path']}/jcr:primaryType`} value="lfs:Conditional" />
+                <TextField
+                  label="Condition JSON"
+                  name={`${data['@path']}`}
+                  defaultValue={isEdit && data.condition ? JSON.stringify(data.condition) : '' }
+                  style={{width: "500px"}}
+                  multiline
+                />
+              </React.Fragment>
```
Adding this here seems to trigger the inability to save other section properties

marta- committed:
LFS-654: Better UI for Questionnaire wizard
Displayed all available section properties in view mode

Commit:
LFS-654: Better UI for Questionnaire wizard
Removed dependency ment to be in LFS-603

veronikaslc committed:
LFS-654: Better UI for Questionnaire wizard

Small typos fix

Commit:
LFS-654: Better UI for Questionnaire wizard
Small UI improvements; Fixed not saving section details error;

Commit:
Merge branch 'LFS-654' of https://github.com/ccmbioinfo/lfs into LFS-654

Commit:
LFS-654: LFS-654: Better UI for Questionnaire wizard
Fixed React warning
Improoved style code

Commit:
LFS-654: Better UI for Questionnaire wizard
Added CreationMenu component to reduce code duplication
Removed condition from edit mode

Commit:
LFS-654: Better UI for Questionnaire wizard
Better column spasing for edit dialog

Commit:
LFS-673: The questionnaire wizard fails to show updated questionnaire requiredSubjectTypes property that was previously unset
Adjust code to consider requiredSubjectTypes to be an object

Commit:
LFS-654: Better UI for Questionnaire wizard
Further refactoring to avoid code duplication

veronikaslc commented:
@sashaandjic Ready for test

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/QuestionnaireItemCard.jsx:
```diff
@@ -0,0 +1,97 @@
+  Avatar,
+  Card,
+  CardContent,
+  CardHeader,
```
We don't need to import CardHeader here, we're only using QuestionnaireCardHeader

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/QuestionnaireCardHeader.jsx:
```diff
@@ -0,0 +1,54 @@
+
+let QuestionnaireCardHeader = (props) => {
+  return (
+    <>
```
It's not necessary to wrap the CardHeader in a React fragment

Commit:
LFS-654: Better UI for Questionnaire wizard
Better spacing for edit dialog

Commit:
LFS-686: In the questionnaire wizard, the last predefined answer option provided for a question does not get saved

marta- committed:
Merge branch 'dev' into LFS-654

marta- committed:
LFS-643: Misplaced Data Type label in Admin's Edit Question dialog
Fixed. Also removed excessive indentation of data type specific properties.

sashaandjic commented:
The following problems are fixed: 
- LFS-643
- LFS-686
 Problem descripbed in LFS-673 is fixed, however the new problem is observed. Steps to reproduce:
- run LFS in cardiac mode
- go to Administration -> Questionnaries -> any cardiac questionnaire that doesn't have the `requiredSubjectTypes` property already set
- edit `requiredSubjectTypes` and save
- open it in the edit mode -> the screen will go blank
![image](https://user-images.githubusercontent.com/23243244/98312589-0d1a7c00-1fa0-11eb-93bb-0c532b514d31.png)
Please note that editing the **Maximum forms of this type per subject** field has no affect. It can be changed or left blank.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -98,15 +98,24 @@ let AnswerOptions = (props) => {
+    // Have to manually invoke submit with timeout to let re-rendering of adding new answer option complete
+    // Cause: Calling onBlur and mutating state can cause onClick for form submit to not fire
+    // Issue details: https://github.com/facebook/react/issues/4210
+    if (event?.relatedTarget.type == "submit") {
```
This occasionally causes a console error because relatedTarget can be undefined.
```suggestion
    if (event?.relatedTarget?.type == "submit") {
```

marta- committed:
LFS-654: Better UI for Questionnaire wizard
Improved the look of the edit dialog (field labels, spacing) and the question properties in view mode

marta- committed:
LFS-686: In the questionnaire wizard, the last predefined answer option provided for a question does not get saved
Fixed console error

LFS-654: Better UI for Questionnaire wizard
Minor code cleanup

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/QuestionnaireCardHeader.jsx:
```diff
@@ -0,0 +1,54 @@
+        title={
+          <div>
+            <Typography variant="overline">{props.type}</Typography>
+            <Typography variant="h6">{props.label}</Typography>
```
This introduces a regression, Question text is no longer displayed in the title bar

sashaandjic commented:
LFS-643 fixed.

marta- committed:
LFS-673:The questionnaire wizard fails to show updated questionnaire requiredSubjectTypes property that was previously unset
Fixed bug with editing a requiredSubjectTypes field with a single value, which was passed as an object instead of an array

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -297,6 +297,28 @@ const questionnaireStyle = theme => ({
+        border: "0 none",
+        paddingLeft: 0,
+        paddingTop: 0,
+        paddingBottom: 0,
```
Padding bottom shouldn't be set to 0, it makes the grid look too crowded

marta- committed:
LFS-654: Better UI for Questionnaire wizard
* displayed the Question text in the Question card again
* improved the spacing in the questionnaire / section properties grid
* made the title input full width in the edit dialog

marta- committed:
LFS-654: Minor improvements to the delete dialog

marta- committed:
LFS-654: Better UI for Questionnaire wizard
[refactoring] Used the QuestionnaireItemCard for questionnaire properties; tidied up styling

sashaandjic commented:
LFS-673, LFS-686, LFS-643 are fixed now.

marta- committed:
LFS-654: Better UI for Questionnaire wizard
[refactoring] Renamed function that gets called after a questionnaire item action is done

sashaandjic commented:
Where did the title go? It is available only when creating new question/section.
![image](https://user-images.githubusercontent.com/23243244/98406388-dd20b680-203b-11eb-982f-dba80ea1866e.png)
![image](https://user-images.githubusercontent.com/23243244/98406418-ed389600-203b-11eb-9f9b-d74267ee94e6.png)

sashaandjic commented:
Can we remove #1 if section is defined as non-recurrent?
For recurrent sections I suggest putting #1 of #total number of reccurences?
![image](https://user-images.githubusercontent.com/23243244/98407017-efe7bb00-203c-11eb-97ee-bdc3ab3f7134.png)
![image](https://user-images.githubusercontent.com/23243244/98407039-f83ff600-203c-11eb-8ead-8b977b9b5ec8.png)

sashaandjic commented:
How do I make a Boolean field required?

marta- commented:
> Can we remove #1 if section is defined as non-recurrent?
> For recurrent sections I suggest putting #1 of #total number of recurrences?

Please report as a separate tasks, these improvements are not in scope.

marta- commented:
> Where did the title go? It is available only when creating new question/section.

Unfortunately `title` is a very misleading label for that field. It's actually the JCR node name, and will be part of the URL where the question will be stored. It's also the variable name. It cannot be changed because the question can no longer be moved in edit mode. I'll clarify in the UI.

marta- committed:
LFS-654: Better UI for Questionnaire wizard
Refactoring:
* New reusable component to display questionnaire contents and section contents consistently
* Replaced custom styling with adjusted grid spacing

sashaandjic commented:
> Unfortunately `title` is a very misleading label for that field.

Should we call it ID?

marta- committed:
LFS-654: Better UI for Questionnaire wizard
Renamed misleading "title" field to "Variable name" for questions and "Section identifier" for sections. Also displayed it read-only in edit mode.

marta- committed:
LFS-654: Better UI for Questionnaire wizard
Allow to make boolean questions mandatory.

sashaandjic commented:
tested; all of my complaints have been resolved

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -98,15 +98,24 @@ let AnswerOptions = (props) => {
+    // Issue details: https://github.com/facebook/react/issues/4210
+    if (event?.relatedTarget?.type == "submit") {
+      const timer = setTimeout(() => {
+        saveButtonRef?.current?.click();
```
Misbehavior:
- edit a question
- type an answer option
- press Tab
- the focus switches from the answer option input to the Save button, and the dialog closes

It's not a big deal, it can even be considered a feature that you can just tab your way through the form, but it's still unexpected that Tab closes the dialog.

veronikaslc replied:
To be fixed in LFS-700

sdumitriu reviewed (approved):
Some things:
- `minAnswers` and `maxAnswers` are created as `String` properties, they should be `Long` instead
- there's a `title` property created for new questions that shouldn't be there
- the `display` property is not exposed, so it is not possible to create a new "list only" or a "textarea" question (and probably others)

However, this is already an impressive improvement.

marta- commented:
> * there's a `title` property created for new questions that shouldn't be there
> * the `display` property is not exposed, so it is not possible to create a new "list only" or a "textarea" question (and probably others)

Addressed in the last commit.

marta- committed:
LFS-654: Better UI for Questionnaire wizard
* Enabled displayMode
* Prevent creating a useless "title" property

Not done yet: only display "Answer options" if displayMode has "list" in it.

marta- committed:
LFS-654: Better UI for Questionnaire wizard
Better management pf min/max answers for boolean questions in the edit mode

marta- commented:
> * `minAnswers` and `maxAnswers` are created as `String` properties, they should be `Long` instead

This seems to happen only for boolean questions. Also, maxAnswers should not even be available for booleans. I committed a quick fix which should cover that.

marta- committed:
Merge branch 'dev' into LFS-654

OrdiNeu committed:
LFS-654: Better UI for Questionnaire wizard
Remove unused imports

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -149,129 +98,105 @@ let Questionnaire = (props) => {
-            open={Boolean(anchorEl)}
-            onClose={handleCloseMenu}
-            disableAutoFocusItem
+          <QuestionnaireItemCard
```
Can't figure out why, but something's wrong with the "Questionnaire properties" item card, where if you open the edit menu once, close it, and then try to reopen it, it won't reopen. I'm not sure what's causing this, but all of the other item cards work fine.

marta- replied:
@OrdiNeu This should be fixed now

OrdiNeu reviewed (approved):
Looks much better than before!

marta- committed:
LFS-654: Better UI for Questionnaire wizard
Fixed question card avatar

marta- committed:
LFS-654: Better UI for Questionnaire wizard
Fixed data loading glitch and questionnaire properties editing bug
